### PR TITLE
[SPARK-39936][SQL] Store schema in properties for Spark Views

### PR DESCRIPTION
**Pull request by Jeffreychen99**
### What changes were proposed in this pull request?

We normally store the tableSchema in table properties rather than as an actual schema. We normally do this as it helps bypass some hive-metastore issues including the parsing with the SparkSQLDataType parser.

However, the issue is that we were not emptying out the tableSchema for spark views (which are incompatible with Hive to start with). Thus, the hive DataType parser would try to parse the schema and it would break. The fix is to empty out the tableSchema for spark views when saving them to Hive Metastore.

### Why are the changes needed?

This fixes the following bug:
```
-- this works since we use backticks in `date-of-creation`
create table table_with_hyphen (f array<struct<validColumnTypeName:string>>) using parquet;
-- this should work since we use backticks in `date-of-creation` but will break without this bugfix
create or replace view view_with_hyphen as select f from table_with_hyphen;
```

### Does this PR introduce _any_ user-facing change?

Bugfix as described above.

### How was this patch tested?

Tested in spark and added a unit test.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -283,7 +283,22 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
         // columns etc. in table properties, so that we can work around the Hive metastore issue
         // about not case preserving and make Hive serde table and view support mixed-case column
         // names.
-        properties = tableDefinition.properties ++ tableMetaToTableProps(tableDefinition))
+        properties = tableDefinition.properties ++ tableMetaToTableProps(tableDefinition)
+      )
+      try {
+        client.createTable(tableWithDataSourceProps, ignoreIfExists)
+      } catch {
+        case NonFatal(e) =>
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:291]:
```suggestion
        case NonFatal(e) if ...is_view... =>
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 2]
```diff
@@ -283,7 +283,22 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
...
+          // If for some reason we fail to store the schema we store it as empty there
+          // since we already store the real schema in the table properties. This try-catch
+          // should only be necessary for Spark views which are incompatible with Hive
+          client.createTable(
+            tableWithDataSourceProps.copy(
+              schema = StructType(EMPTY_DATA_SCHEMA ++ tableDefinition.partitionSchema)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:297]: A view cannot have partition columns, so here can simply be `new StructType()`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 3]
```diff
@@ -160,18 +160,6 @@ class HiveCatalogedDDLSuite extends DDLSuite with TestHiveSingleton with BeforeA
   }
 
   test("SPARK-22431: illegal nested type") {
-    val queries = Seq(
-      "CREATE TABLE t USING hive AS SELECT STRUCT('a' AS `$a`, 1 AS b) q",
-      "CREATE TABLE t(q STRUCT<`$a`:INT, col2:STRING>, i1 INT) USING hive",
-      "CREATE VIEW t AS SELECT STRUCT('a' AS `$a`, 1 AS b) q")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:166]: I think we only need to remove this line from the test

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 4]
```diff
@@ -378,4 +378,15 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest with ParquetTest {
...
   }
+
+  test("Create view with dashes in column type") {
+    sql(
+      """
+        |CREATE TABLE tbl (f array<struct<`date-of-creation`:string>>) USING parquet
+        |""".stripMargin)
+    sql(
+      """
+        |CREATE VIEW tblView AS SELECT f FROM tbl
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala:389]: we can simply use the existing test case `CREATE VIEW t AS SELECT STRUCT('a' AS `$a`, 1 AS b) q`

Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala:389]: and let's make the test complete
```
withView("t") {
  sql("CREATE VIEW t AS SELECT STRUCT('a' AS $a, 1 AS b) q")
  checkAnswer(spark.table("t"), Row(Row("a", 1)))
}
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 5]
```diff
@@ -2386,10 +2385,6 @@ class HiveDDLSuite
         "CREATE TABLE t1 USING PARQUET AS SELECT NULL AS null_col",
         "Parquet data source does not support void data type")
 
-      assertAnalysisError(
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:2389]: do we still need to remove this test?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 6]
```diff
@@ -2400,10 +2395,6 @@ class HiveDDLSuite
         "CREATE TABLE t1 (v VOID) USING PARQUET",
         "Parquet data source does not support void data type")
 
-      assertAnalysisError(
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:2403]: ditto

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 5 above)
Comment (Jeffreychen99) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Oops - sorry forgot to add them back

**Review by cloud-fan — APPROVED** [review]
LGTM if tests pass (I think there are still some code style issues which will be caught by the checkers)

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**